### PR TITLE
Kernel hint design

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: doc/design/kernel_hint_design.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+### final choice
+To provide enough flexibility while avoiding confusion definition, we can predefine some options, such as `force_cpu`, `use_cudnn`, `use_mkldnn` for a user to choose.
+
+```cpp
+const std::string kNonHint = "";
+const std::string kForceCPU = "force_cpu";
+const std::string kUseCUDNN = "use_cudnn";
+const std::string kUseMKLDNN = "use_mkldnn";
+
+KernelType GetExpectedKernelTyp() {
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: Typo

File: doc/design/kernel_hint_design.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+In the current design,  we use KernelType to describe one kernel.
+
+```cpp
+struct KernelType {
+  Place place_;
+  DataType data_type_;
+  LayoutType layout_;
+};
+```
+ `place_` `data_type_` and `layout_` can come from the input tensor of the operator, `GetActualKernelType(inputs)` use inputs to infer the proper kernel key that fit the incoming data, user can not config it.
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: `Users cannot configure it.`

Actually, users can configure input data and influence the `actual kernel type`.

File: doc/design/kernel_hint_design.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+The design also provides a virtual method `GetExpectedKernelType` that user can overload and choose the KernelType they want to use.
+
+so, we should send the information user defined in proto to `GetExpectedKernelType` for choosing a kernel.
+
+The problem is, how should we define and send the information for `GetExpectedKernelType` to use?
+
+## Solution
+
+### potential choice
+1, Do nothing, let the user add the information they want to operator‘s attribute and get them inside `GetExpectedKernelType`, this can work right. But there is a little problem that users may define many kinds of hints for the same purpose, such as `force_cpu`, `use_cpu`, `CPU` for CPU kernel, and `use_cudnn`, `force_cudnn`, `cudnn_kernel` for use of CUDNN kernel.
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: `1,` --> `1.`

Comment (reyoung) [inline doc/design/kernel_hint_design.md]: We can define some global constant for these attribute name.

File: doc/design/kernel_hint_design.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+2, Pre-define all the needed option and use a single attr key such as `kernel_hint` for the user, this is not so flexible if the user wants to define some more kind of hint.
+
+### final choice
+To provide enough flexibility while avoiding confusion definition, we can predefine some options, such as `force_cpu`, `use_cudnn`, `use_mkldnn` for a user to choose.
+
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: Using string to represent `force_cpu`, `use_cudnn`, `use_mkldnn` is not extendable. How to make an operator use CPU forcibly and use MKLDNN?

File: doc/design/kernel_hint_design.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+```cpp
+const std::string kNonHint = "";
+const std::string kForceCPU = "force_cpu";
+const std::string kUseCUDNN = "use_cudnn";
+const std::string kUseMKLDNN = "use_mkldnn";
+
+KernelType GetExpectedKernelTyp() {
+    // "kernel_hint" is a user defined attribute name
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: I do not think `kernel_hint` is a formal name. It is confusing for users.

File: doc/design/kernel_hint_design.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	if (Attr<std::string>("kernel_hint") == kForceCPU) {
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: Why not use `Attr<bool>(kForceCPU)` ?

File: doc/design/kernel_hint_design.md (hunk 1 above)
Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: fixed

File: doc/design/kernel_hint_design.md (hunk 2 above)
Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: done

File: doc/design/kernel_hint_design.md (hunk 3 above)
Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: done

Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: done

File: doc/design/kernel_hint_design.md (hunk 5 above)
Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: done, use force_cpu as key and bool as value.

File: doc/design/kernel_hint_design.md (hunk 6 above)
Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: done

File: doc/design/kernel_hint_design.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+In C++
+
+```cpp
+const std::string kForceCPU = "force_cpu";
+const std::string kUseCUDNN = "use_cudnn";
+const std::string kUseMKLDNN = "use_mkldnn";
+
+KernelType GetExpectedKernelType() {
+	if (Attr<bool>(kForceCPU)) {
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: Need a better Indent

File: doc/design/kernel_hint_design.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+}
+```
+
+In Python code
+
+```python
+FORCE_CPU = core.kForceCPU()
+
+def xx_layer(..., force_cpu=false):
+	layer_helper = LayerHelper(...)
```
Comment (reyoung) [inline doc/design/kernel_hint_design.md]: need a better indent

Comment (jacquesqiao) [reply doc/design/kernel_hint_design.md]: change tab to space

**Pull request merged**